controllers/network/openstackipset_controller.go

```diff
@@ -0,0 +1,264 @@
+//+kubebuilder:rbac:groups=network.openstack.org,resources=openstacknets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=network.openstack.org,resources=openstackipsets/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=network.openstack.org,resources=openstackipsets/finalizers,verbs=update
+//+kubebuilder:rbac:groups=network.openstack.org,resources=openstacknets/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the OpenStackIPSet object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (r *OpenStackIPSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, _err error) {
+	_ = log.FromContext(ctx)
+	result = ctrl.Result{}
+	// TODO(user): your logic here
+	r.Log.Info("started reconciliation")
+
+	instance := &networkv1beta1.OpenStackIPSet{}
+	err := r.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers. Return and don't requeue.
+			return result, nil
+		}
+		return result, err
+	}
+
+	helper, err := helper.NewHelper(
+		instance,
+		r.Client,
+		r.Kclient,
+		r.Scheme,
+		r.Log,
+	)
+
+	if err != nil {
+		// helper might be nil, so can't use util.LogErrorForObject since it requires helper as first arg
+		r.Log.Error(err, fmt.Sprintf("unable to acquire helper for OpenStackIPSet %s", instance.Name))
+		return result, err
+	}
+
+	defer func() {
+		if err != nil {
+			instance.Status.Conditions.MarkTrue(networkv1beta1.OpenStackIPSetReadyCondition, networkv1beta1.OpenStackIPSetReadyMessage)
+		}
+		// update the overall status condition if service is ready
+		err := helper.PatchInstance(ctx, instance)
+		if err != nil {
+			r.Log.Error(err, "Can't patch Instance")
+			_err = err
+			return
+		}
+	}()
+
+	if instance.DeletionTimestamp.IsZero() && controllerutil.AddFinalizer(instance, helper.GetFinalizer()) {
+		return ctrl.Result{}, err
+	}
+
+	if !instance.DeletionTimestamp.IsZero() {
+		return r.ReconcileDelete(ctx, instance, helper, req)
+	}
+
+	instance.InitCondition()
+
+	// Build AssignIPDetails
+	networks := instance.Spec.Networks
+
+	if instance.Status.IPAddresses == nil {
+		instance.Status.IPAddresses = make(map[string]networkv1beta1.IPReservation)
+	}
+
+	//find the the openstacknet objects
+	for _, network := range networks {
+		if _, ok := instance.Status.IPAddresses[network.Name]; ok {
```

Comment:
should we check if the reservation is still on the osnet, or is it save to say we stored it once we expect it to be there? if we change the reservation to be a map[string]IPReservation, this would be easy to check

Reply:
We can surely add more checks for cases like that, but I did not want to make the CRDs overly complex. Like OpenStackNet to know who has requested/reserved it.

Reply:
we did it like that in ospdo and just marked remove nodes/entries as deleted so that you could replace a node and get the same reservation. we had a dedicated parameter to always clean reservations on delete. but we could also say if you want a behavior like that, always request a specific IP in the request. either way works